Lombiq.Tests.UI/Services/UITestExecutionSession.cs

```diff
@@ -98,7 +105,30 @@ public async Task<bool> ExecuteAsync(int retryCount, string dumpRootPath)
 
             if (_context.IsBrowserConfigured) _context.SetDefaultBrowserSize();
 
-            await _testManifest.TestAsync(_context);
+            var timeout = _configuration.TimeoutConfiguration.TestRunTimeout;
+
+            var timeoutTask = Task.Delay(timeout, _configuration.TestCancellationToken);
+            var testTask = _testManifest.TestAsync(_context);
+
+            await Task.WhenAny(testTask, timeoutTask);
+
+            if (timeoutTask.IsCompleted && !_configuration.TestCancellationToken.IsCancellationRequested)
+            {
+                // If the EnterInteractiveModeAsync() extension method has been used, then timeout should be ignored to
+                // make the debugging experience smoother. Note that EnterInteractiveModeAsync() should never be used in
+                // committed tests.
+                if (!ShortcutsUITestContextExtensions.InteractiveModeHasBeenUsed)
+                {
+                    throw new TimeoutException($"The time allotted for the test ({timeout}) was exceeded.");
+                }
+
+                await testTask;
+            }
+
+            // Since the timeout task is not yet completed but the Task.WhenAny has finished, the test task is done in
+            // some way. So it's safe to await it here. It's also necessary to cleanly propagate any exceptions that may
+            // have been thrown inside it.
+            await testTask;
```

Comment:
I think we dont need to `await` twice
```suggestion
            }

            // Since the timeout task is not yet completed but the Task.WhenAny has finished, the test task is done in
            // some way. So it's safe to await it here. It's also necessary to cleanly propagate any exceptions that may
            // have been thrown inside it.
            await testTask;
```

Reply:
Indeed.